File: frontend/src/api/tauri/pdf.ts
/** PDF processing — classify, extract, parse, full pipeline. */

import { invoke } from '@tauri-apps/api/core'
import { invokeWithError } from './_utils'
import { ErrorCode } from '../../utils/errors'
import type { ActivityData, DocumentClassification } from './text'

// ---- pdf-inspector ----

export interface PdfClassification {
  pdf_type: string
  confidence: number
  page_count: number
  pages_needing_ocr: number[]
  text_density_avg: number
  has_complex_layout: boolean
  has_encoding_issues: boolean
  title: string | null
}

export interface PdfExtraction {
  markdown: string
  page_count: number
  pages_needing_ocr: number[]
  confidence: number
  has_complex_layout: boolean
  has_encoding_issues: boolean
}

export async function classifyPdf(path: string): Promise<PdfClassification> {
  return invokeWithError(
    () => invoke<PdfClassification>('classify_pdf', { path }),
    ErrorCode.PdfParse,
  )
}

export async function inspectPdf(
  projectRoot: string,
  docId: string,
): Promise<PdfClassification> {
  return invokeWithError(
    () => invoke<PdfClassification>('inspect_pdf', { projectRoot, docId }),
    ErrorCode.PdfParse,
  )
}

export async function confirmOcr(
  projectRoot: string,
  docId: string,
  confirm: boolean,
): Promise<{ success: boolean; doc_id: string; ocr_status: string; task_id: string }> {
  return invokeWithError(
    () => invoke<{ success: boolean; doc_id: string; ocr_status: string; task_id: string }>('confirm_ocr', {
      projectRoot,
      docId,
      confirm,
    }),
    ErrorCode.PdfParse,
  )
}

export async function extractText(path: string): Promise<PdfExtraction> {
  return invokeWithError(
    () => invoke<PdfExtraction>('extract_text', { path }),
    ErrorCode.PdfParse,
  )
}

// ---- pipeline ----

export interface ImageRef {
  filename: string
  page: number
  region: string | null
  description: string | null
  esmiles: string | null
  rel_path: string | null
}

export interface Heading {
  level: number
  title: string
  line_num: number
}

export interface SectionChunk {
  title: string
  path: string
  text: string
  page_start: number | null
  page_end: number | null
  line_start: number
  line_end: number
}

export interface PdfParseResult {
  content: string
  classification: DocumentClassification
  chunks: string[]
  esmiles: string[]
  activities: ActivityData[]
  parser: string
  page_count: number
  images: ImageRef[]
  headings: Heading[]
  sections: SectionChunk[]
  page_texts: string[]
}

export async function parsePdf(
  path: string,
  chunkSize?: number,
  overlap?: number,
  parser?: string,
): Promise<PdfParseResult> {
  return invokeWithError(
    () => invoke<PdfParseResult>('parse_pdf', {
      path,
      chunkSize: chunkSize ?? 512,
      overlap: overlap ?? 128,
      parser: parser ?? 'pdf_inspector',
    }),
    ErrorCode.PdfParse,
  )
}

// ---- process_document (A3: 完整文档处理管线) ----

export interface DocProgressEvent {
  stage: string
  payload: Record<string, unknown>
}

export async function processDocument(
  path: string,
  userRequest?: string,
  projectRoot?: string,
): Promise<void> {
  return invokeWithError(
    () => invoke('process_document', {
      path,
      userRequest: userRequest ?? '',
      projectRoot,
    }),
    ErrorCode.PdfParse,
  )
}

// ---- OCR 布局可视化 ----

export interface OcrBlock {
  page: number
  block_type: string
  bbox: [number, number, number, number]
  content: string | null
  index: number
  angle: number
}

export interface OcrLayoutResult {
  path: string
  parser: string
  page_count: number
  blocks: OcrBlock[]
  from_cache: boolean
}

export async function getDocumentOcrLayout(path: string, doc_id?: string): Promise<OcrLayoutResult> {
  return invokeWithError(
    () => invoke<OcrLayoutResult>('get_document_ocr_layout', { path, doc_id }),
    ErrorCode.PdfParse,
  )
}

// ---- 嵌入图在 PDF 页面上的 bbox（coref 投影用） ----

export interface FigureBbox {
  xref: number
  /** [x1, y1, x2, y2] in PDF points (bottom-left origin) */
  bbox_pdf: [number, number, number, number]
  width: number | null
  height: number | null
}

export interface PageFigureBboxes {
  page_num: number
  figures: FigureBbox[]
}

/** 返回整本 PDF 中所有嵌入图在各自页面上的 bbox。
 *
 * 用于 coref overlay 把 figure 局部 0-1 坐标投影到 PDF 页面坐标。
 */
export async function getFigureBboxes(pdfPath: string): Promise<PageFigureBboxes[]> {
  return invokeWithError(
    () => invoke<PageFigureBboxes[]>('get_figure_bboxes', { pdfPath }),
    ErrorCode.PdfParse,
  )
}

File: src/mbforge/routers/pdf.py
"""PDF processing endpoints — classification + inspection stubs.

Full PDF parsing is implemented by `pipeline/extract_text.py` and
`routers/moldet_api.py:extract_pdf_page`. The endpoints here cover the
frontend's lightweight probes (classify, ocr-layout, figure-bboxes) used
by the queue UI before kicking off a full pipeline run.
"""

from __future__ import annotations

from fastapi import APIRouter

router = APIRouter()


@router.post("/classify")
async def classify_pdf(body: dict) -> dict:
    """Classify PDF type stub."""
    return {
        "pdf_type": "unknown",
        "confidence": 0.0,
        "page_count": 0,
        "pages_needing_ocr": [],
        "text_density_avg": 0.0,
        "has_complex_layout": False,
        "has_encoding_issues": False,
        "title": None,
    }


@router.post("/inspect")
async def inspect_pdf(body: dict) -> dict:
    """Inspect PDF stub."""
    return {
        "pdf_type": "unknown",
        "confidence": 0.0,
        "page_count": 0,
        "pages_needing_ocr": [],
        "text_density_avg": 0.0,
        "has_complex_layout": False,
        "has_encoding_issues": False,
        "title": None,
    }


@router.post("/confirm-ocr")
async def confirm_ocr(body: dict) -> dict:
    """Confirm OCR stub."""
    return {"success": True, "doc_id": body.get("docId", ""), "ocr_status": "pending", "task_id": ""}


@router.post("/extract-text")
async def extract_text(body: dict) -> dict:
    """Extract text stub."""
    return {
        "markdown": "",
        "page_count": 0,
        "pages_needing_ocr": [],
        "confidence": 0.0,
        "has_complex_layout": False,
        "has_encoding_issues": False,
    }


@router.post("/parse")
async def parse_pdf(body: dict) -> dict:
    """Full PDF parse stub."""
    return {
        "content": "",
        "classification": {
            "text_density": 0.0,
            "is_scanned": False,
            "has_molecular_patterns": False,
            "metadata_hints": None,
            "pages": [],
            "needs_confirmation": False,
        },
        "chunks": [],
        "esmiles": [],
        "activities": [],
        "parser": "stub",
        "page_count": 0,
        "images": [],
        "headings": [],
        "sections": [],
        "page_texts": [],
    }


@router.post("/process-document")
async def process_document(body: dict) -> dict:
    """Process document stub."""
    return {"success": True}


@router.post("/ocr-layout")
async def ocr_layout(body: dict) -> dict:
    """OCR layout stub."""
    return {"path": body.get("path", ""), "parser": "stub", "page_count": 0, "blocks": [], "from_cache": False}


@router.post("/figure-bboxes")
async def figure_bboxes(body: dict) -> list[dict]:
    """Return per-page figure bbox arrays for coref overlay projection.

    Real implementation lives in the pipeline layer (calls the figure-bbox
    detector). This stub preserves the API contract expected by the
    frontend (``usePdfViewer.ts`` → ``getFigureBboxes`` → ``PageFigureBboxes[]``)
    so the endpoint always parses against its declared return type even when
    the underlying detector is unavailable.

    Contract:
      - ``{"pdf_path": "<abs path>"}`` → list of ``{page_num, figures}``
        entries, one per PDF page. ``figures`` is the bbox array (empty when
        the page has no extractable figures).
      - Missing/unreadable PDF → ``[]``.
    """
    import fitz  # local import keeps the router importable without PyMuPDF

    pdf_path = body.get("pdf_path") or body.get("pdfPath") or ""
    if not pdf_path:
        return []
    try:
        with fitz.open(pdf_path) as doc:
            return [
                {"page_num": idx + 1, "figures": []}
                for idx in range(doc.page_count)
            ]
    except Exception:
        # File missing, encrypted, or unreadable — surface as empty result
        # rather than 500, so the frontend can fall back gracefully.
        return []
